File: app/src/main/java/rx_playground/com/jablonski/rxandroidplayground/modeldetails/ModelDetailsPresenter.java
package rx_playground.com.jablonski.rxandroidplayground.modeldetails;

import java.util.ArrayList;
import java.util.List;

import rx_playground.com.jablonski.rxandroidplayground.model.ImageSource;
import rx_playground.com.jablonski.rxandroidplayground.model.ListElement;
import rx_playground.com.jablonski.rxandroidplayground.model.Model;
import rx_playground.com.jablonski.rxandroidplayground.model.Photo;
import rx_playground.com.jablonski.rxandroidplayground.model.RowConfig;
import rx_playground.com.jablonski.rxandroidplayground.modeldetails.ModelDetailsContract;

/**
 * Created by yabol on 16.05.2017.
 */

public class ModelDetailsPresenter implements ModelDetailsContract.Presenter, ModelDetailsContract.Provider, ModelDetailsContract.ImagesProvider {
    private Model model;
    private List<ListElement> elements;
    private ModelDetailsContract.View view;
    private ModelDetailsContract.Repository repository;
    private Photo photo;

    public ModelDetailsPresenter(ModelDetailsContract.View view){
        this.view = view;
        this.elements = new ArrayList<>();
    }

    public void setRepository(ModelDetailsContract.Repository repository){
        this.repository = repository;
    }

    @Override
    public void loadModelData(String modelId) {
        this.repository.getModelDetails(modelId);
        this.repository.getModelPhotos(modelId);
    }

    @Override
    public void displayModel(Model model) {
        this.model = model;
        if(this.elements.size() > 0){
            this.elements.clear();
        }
        this.elements.add(model);
        this.elements.add(model.getCategory());
        this.elements.add(model.getEngine());
        this.elements.add(model.getTransmission());
        this.view.displayModelDetails();
    }

    @Override
    public void displayPhotos(Photo photos) {
        this.photo = photos;
        this.view.displayPhotos();
    }

    public Photo getPhoto(){
        return this.photo;
    }

    @Override
    public Model getModel() {
        return this.model;
    }

    @Override
    public RowConfig getObject(int position) {
        return this.elements.get(position).getRowConfig();
    }

    @Override
    public int getCount() {
        return this.elements.size();
    }

    @Override
    public int getPositionType(int position) {
        return this.elements.get(position).getViewType();
    }

    @Override
    public ImageSource getImage(int position) {
        if(this.photo != null){
            return this.photo.getSources().get(position);
        }
        return null;
    }

    @Override
    public int getPhotosCount() {
        if(this.photo != null){
            return this.photo.getSources().size();
        }
        return 0;
    }
}

File: app/src/main/java/rx_playground/com/jablonski/rxandroidplayground/modeldetails/ModelDetailsPresenter.kt
package rx_playground.com.jablonski.rxandroidplayground.modeldetails

import android.util.Log
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import rx_playground.com.jablonski.rxandroidplayground.model.*
import java.util.*

/**
 * Created by yabol on 16.05.2017.
 */

class ModelDetailsPresenter(private val view: ModelDetailsContract.View, private val repository: ModelDetailsContract.Repository) :
        ModelDetailsContract.Presenter,
        ModelDetailsContract.Provider,
        ModelDetailsContract.ImagesProvider {

    var model: Model? = null
    private val elements: MutableList<ListElement>
    var photo: Photo? = null
        private set

    override val count: Int
        get() = this.elements.size

    override val photosCount: Int
        get() = if (this.photo != null) {
            this.photo!!.sources.size
        } else 0

    init {
        this.elements = ArrayList()
    }

    override fun loadModelData(modelId: String) {

        this.repository.getModelDetails(modelId).
                subscribeOn(Schedulers.newThread()).
                observeOn(AndroidSchedulers.mainThread()).
                subscribe({
                    val result = it.model
                    result.engine = it.engine
                    result.transmission = it.transmission
                    result.category = it.categories
                    result.drivenWheels = it.drivenWheels
                    result.numberOfDoors = it.numOfDoors

                    displayModel(result)
                }, {
                    Log.e("load model error", it.message)
                })
        this.repository.getModelPhotos(modelId).
                subscribeOn(Schedulers.newThread()).
                observeOn(AndroidSchedulers.mainThread()).
                subscribe({
                    val photos = it.photos
                    val resultPhoto = Photo()
                    for (singlePhoto in photos) {
                        resultPhoto.addPhotoSources(singlePhoto.sources)
                    }
                    displayPhotos(resultPhoto)
                }, {
                    Log.e("Load photos error", it.message)
                })
    }


    override fun displayModel(model: Model) {
        this.model = model
        if (this.elements.size > 0) {
            this.elements.clear()
        }
        with(elements) {
            add(model)
            add(model.category)
            add(model.engine)
            add(model.transmission)
        }
        view.displayModelDetails(model)
    }

    override fun displayPhotos(photos: Photo?) {
        this.photo = photos
        this.view.displayPhotos()
    }

    override fun getObject(position: Int): RowConfig {
        return this.elements[position].rowConfig
    }

    override fun getPositionType(position: Int): Int {
        return this.elements[position].viewType
    }

    override fun getImage(position: Int): ImageSource? {
        return if (this.photo != null) {
            this.photo!!.sources[position]
        } else return null
    }
}
